feat(quasar): add docsUrl to API definition

lucasfernog opened the pull request:
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [x] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch and _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on a Cordova (iOS, Android) app
- [ ] It's been tested on a Electron app
- [x] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar-framework.org/tree/dev/source) (for faster update click on "Suggest an edit on GitHub" at bottom of page) or explained in the PR's description.

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

The ```quasar docs``` command receives a Quasar component/directive/plugin and opens the docs page. It also supports app extensions. It also has a --api option, which uses the API table anchor as URL.

Credits to @hawkeye64 for the idea.

EDIT: added options to open sections (components, plugins, directives and extensions), e.g. ```quasar docs components```

lucasfernog committed:
quasar docs command

lucasfernog committed:
add docs

nothingismagick commented:
Wow. Congrats on this!!! Great job

lucasfernog commented:
Thanks. Will adjust it to support ```quasar docs components/plugins/directives```.

nothingismagick commented:
Just a dumb question: how hard would it be to have

$ quasar api search ${search_term}

lucasfernog committed:
add sections

lucasfernog commented:
> Just a dumb question: how hard would it be to have $ quasar api search ${search_term}

That's a VERY nice idea, and not hard at all. Just a matter of reading all API file names (quasar/dist/api/*.json) and string filtering them. If it will be another command, I can make another PR.

nothingismagick commented:
Oh please, do it.

On Sun, 5 May 2019, 00:13 Lucas Fernandes Nogueira, <
notifications@github.com> wrote:

> Just a dumb question: how hard would it be to have $ quasar api search
> ${search_term}
>
> That's a VERY nice idea, and not hard at all. Just a matter of reading all
> API file names (quasar/dist/api/*.json) and string filtering them. If it
> will be another command, I can make another PR.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/quasarframework/quasar/pull/4045#issuecomment-489369570>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIM4G6AYSEDHOFXRIWKGEWLPTYDAZANCNFSM4HKZYORA>
> .
>

lucasfernog commented:
Another thing that is missing here: can this command read the Quasar build config and get the user preferred browser? I've tried but couldn't do it.

lucasfernog committed:
add sections to command

lucasfernog committed:
fix open browser

lucasfernog committed:
pass supplier to docs function

lucasfernog committed:
read quasar conf for open browser

lucasfernog committed:
fix command with app extensions API

lucasfernog commented:
> Another thing that is missing here: can this command read the Quasar build config and get the user preferred browser? I've tried but couldn't do it.

Fixed with latest commits. Also fixed running with app extension API. This PR is now done unless you want something else.

lucasfernog committed:
open-browser using its own logger

lucasfernog committed:
fix Ripple docs

lucasfernog commented:
TODO: move default "docs" values to the build command.

lucasfernog committed:
move "docs" defaults to build.api.js

lucasfernog committed:
fix DocApi trying to render docs field

lucasfernog committed:
Merge branch 'dev' into feature/docs-command

lucasfernog committed:
Merge branch 'dev' into feature/docs-command

lucasfernog committed:
fix QUploaderAddTrigger page

lucasfernog committed:
Merge remote-tracking branch 'upstream/dev' into feature/docs-command

lucasfernog committed:
feat(api) change "docs" key to "meta"

lucasfernog committed:
fix(api) definitions for meta

lucasfernog committed:
fix(api) definitions for meta & `quasar docs` usage

lucasfernog committed:
feat(api) move docs opening to describe command and prefix API with "docs"

lucasfernog committed:
feat(docs) add new quasar describe option

rstoenescu commented on app/bin/quasar:
```diff
@@ -9,6 +9,7 @@ const commands = [
   'clean',
   'inspect',
   'describe',
+  'docs',
```
now obsolete

rstoenescu commented on app/bin/quasar-describe:
```diff
@@ -522,6 +461,9 @@ function printQuasarConfOptions ({ quasarConfOptions }) {
 function describe (api) {
+  if (apiParts.docs) {
+    openBrowser(api.meta.docsUrl)
+  }
```
Would be good to move this after the switch(). In case it fails it should at least print out the API.

rstoenescu commented on docs/src/components/DocApi.vue:
```diff
@@ -198,7 +198,7 @@ export default {
 
   methods: {
-    parseJson (name, { type, behavior, ...api }) {
+    parseJson (name, { type, behavior, docs, ...api }) {
```
It should be "meta" instead of "docs"

rstoenescu commented on docs/src/pages/quasar-cli/cli-documentation/commands-list.md:
````diff
@@ -436,6 +439,26 @@ $ quasar describe QIcon
+  --api, -a    Opens the API table URL
+  --help, -h   Displays this message
+```
+
````
Now obsolete. To remove.

rstoenescu commented on ui/build/build.api.js:
```diff
@@ -117,6 +118,11 @@ const objectTypes = {
   },
 
+  meta: {
+    props: [ 'docsRoute', 'docsPage', 'docsApiAnchor' ],
```
Let's use one prop only "docsUrl" with full URL. For now, point to `https://v1.quasar.dev`. Example:

```js
meta: {
  docsUrl: 'https://v1.quasar.dev'
}
```

rstoenescu commented on ui/src/components/btn/QBtn.json:
```diff
@@ -15,7 +19,7 @@
       "type": "Number",
       "desc": "Percentage (0.0 < x < 100.0); To be used along 'loading' prop; Display a progress bar on the background",
-      "examples": [ 23 ],
+      "examples": [23],
```
More space around is generally better for the eyes. Not important though, but I see this change everywhere.

rstoenescu commented on ui/src/components/pagination/QPagination.json:
```diff
@@ -79,7 +79,7 @@
       "desc": "Generate link for page buttons; For best performance, reference it from your scope and do not define it inline",
       "params": {
-        "page": {
+        "docsPage": {
```
Find & replace failing :)

rstoenescu commented on ui/src/components/table/QTable.json:
```diff
@@ -404,7 +404,7 @@
           "desc": "Is sorting in descending order?"
         },
-        "page": {
+        "docsPage": {
```
Find & replace failing :)
Same for all others here.

rstoenescu commented on app/bin/quasar-describe:
```diff
@@ -522,6 +461,9 @@ function printQuasarConfOptions ({ quasarConfOptions }) {
 }
 
 function describe (api) {
+  if (apiParts.docs) {
```
"apiParts.docs" ?
Also, make it fail-safe. If it doesn't have the meta.docsUrl, then don't try to open it. The API files with the meta field will only be available in upcoming quasar v1.5.0+ only.

lucasfernog committed:
fix(quasar) code review updates

lucasfernog committed:
fix(app) describe command

rstoenescu committed:
Merge branch 'dev' into feature/docs-command
